cloudinit/config/cc_ansible.py

```diff
@@ -0,0 +1,186 @@
+        repository at the remote URL.
+        """
+    ),
+    "examples": [
+        dedent(
+            """\
+            #cloud-config
+            ansible:
+              install-method: distro
+              pull:
+                url: "https://github.com/holmanb/vmboot.git"
+                playbook-name: ubuntu.yml
+            """
+        ),
+        dedent(
+            """\
+            #cloud-config
+            ansible:
+              package-name: ansible-core
+              install-method: pip
+              pull:
+                url: "https://github.com/holmanb/vmboot.git"
+                playbook-name: ubuntu.yml
+            """
+        ),
+    ],
+}
+
+__doc__ = get_meta_doc(meta)
+
+
+class AnsiblePull(abc.ABC):
+    cmd_version: list = []
+    cmd_pull: list = []
+    env: dict = os.environ.copy()
+
+    def get_version(self) -> Optional[Version]:
+        stdout, _ = subp(self.cmd_version, env=self.env)
+        first_line = stdout.splitlines().pop(0)
+        matches = re.search(r"([\d\.]+)", first_line)
+        if matches:
+            version = matches.group(0)
+            return Version.from_str(version)
+        return None
+
+    def pull(self, *args) -> str:
+        stdout, _ = subp([*self.cmd_pull, *args], env=self.env)
+        return stdout
+
+    def check_deps(self):
+        if not self.is_installed():
+            raise ValueError("command: ansible is not installed")
+
+    @abc.abstractmethod
+    def is_installed(self):
+        pass
+
+    @abc.abstractmethod
+    def install(self, pkg_name: str):
+        pass
+
+
+class AnsiblePullPip(AnsiblePull):
+    def __init__(self):
+        self.cmd_pull = ["ansible-pull"]
+        self.cmd_version = ["ansible-pull", "--version"]
+        self.env["PATH"] = ":".join([self.env["PATH"], "/root/.local/bin/"])
+
+    def install(self, pkg_name: str):
+        """should cloud-init grow an interface for non-distro package
+        managers? this seems reusable
+        """
+        if not self.is_installed():
+            subp(["python3", "-m", "pip", "install", "--user", pkg_name])
+
+    def is_installed(self) -> bool:
+        stdout, _ = subp(["python3", "-m", "pip", "list"])
+        return "ansible" in stdout
+
+
+class AnsiblePullDistro(AnsiblePull):
+    def __init__(self, distro):
+        self.cmd_pull = ["ansible-pull"]
+        self.cmd_version = ["ansible-pull", "--version"]
+        self.distro = distro
+
+    def install(self, pkg_name: str):
+        if not self.is_installed():
+            self.distro.install_packages(pkg_name)
+
+    def is_installed(self) -> bool:
+        return bool(which("ansible"))
+
+
+def handle(name: str, cfg: dict, cloud: Cloud, log: Logger, _):
+    ansible_cfg: dict = cfg.get("ansible", {})
+    if ansible_cfg:
+        validate_config(ansible_cfg)
+        install = ansible_cfg["install-method"]
+        pull_cfg = ansible_cfg.get("pull")
+        if pull_cfg:
+            if install == "pip":
```

Comment:
Nit: I think it would be better to type `ansible` for both branches as following, in order to not tie it to a `AnsiblePull`'s implementation:

```diff
diff --git a/cloudinit/config/cc_ansible.py b/cloudinit/config/cc_ansible.py
index a21a91e8..934c6538 100644
--- a/cloudinit/config/cc_ansible.py
+++ b/cloudinit/config/cc_ansible.py
@@ -132,8 +132,9 @@ def handle(name: str, cfg: dict, cloud: Cloud, log: Logger, _):
         install = ansible_cfg["install-method"]
         pull_cfg = ansible_cfg.get("pull")
         if pull_cfg:
+            ansible: AnsiblePull
             if install == "pip":
-                ansible: AnsiblePull = AnsiblePullPip()
+                ansible = AnsiblePullPip()
             else:
                 ansible = AnsiblePullDistro(cloud.distro)
             ansible.install(ansible_cfg["package-name"])

```